e2e/tests/core/client_test.go

```diff
@@ -119,3 +138,207 @@ func (s *ClientTestSuite) TestClientUpdateProposal_Succeeds() {
 		})
 	})
 }
+
+func (s *ClientTestSuite) TestClient_Update_Misbehaviour() {
+	t := s.T()
+	ctx := context.TODO()
+
+	var (
+		trustedHeight   clienttypes.Height
+		latestHeight    clienttypes.Height
+		clientState     ibcexported.ClientState
+		block           *tmproto.Block
+		signers         []tmtypes.PrivValidator
+		validatorSet    []*tmtypes.Validator
+		maliciousHeader *ibctm.Header
+		err             error
+	)
+
+	relayer, _ := s.SetupChainsRelayerAndChannel(ctx)
+	chainA, chainB := s.GetChains()
+
+	s.Require().NoError(test.WaitForBlocks(ctx, 10, chainA, chainB))
+
+	t.Run("update clients", func(t *testing.T) {
+		err := relayer.UpdateClients(ctx, s.GetRelayerExecReporter(), s.GetPathName(0))
+		s.Require().NoError(err)
+
+		clientState, err = s.QueryClientState(ctx, chainA, ibctesting.FirstClientID)
+		s.Require().NoError(err)
+	})
+
+	t.Run("fetch trusted height", func(t *testing.T) {
+		tmClientState, ok := clientState.(*ibctm.ClientState)
+		s.Require().True(ok)
+
+		trustedHeight, ok = tmClientState.GetLatestHeight().(clienttypes.Height)
+		s.Require().True(ok)
+	})
+
+	t.Run("update clients", func(t *testing.T) {
+		err := relayer.UpdateClients(ctx, s.GetRelayerExecReporter(), s.GetPathName(0))
+		s.Require().NoError(err)
+
+		clientState, err = s.QueryClientState(ctx, chainA, ibctesting.FirstClientID)
+		s.Require().NoError(err)
+	})
+
+	t.Run("fetch client state latest height", func(t *testing.T) {
+		tmClientState, ok := clientState.(*ibctm.ClientState)
+		s.Require().True(ok)
+
+		latestHeight, ok = tmClientState.GetLatestHeight().(clienttypes.Height)
+		s.Require().True(ok)
+	})
+
+	t.Run("create validator set", func(t *testing.T) {
+		var validators []*tmservice.Validator
+
+		t.Run("fetch block at latest client state height", func(t *testing.T) {
+			block, err = s.GetBlockByHeight(ctx, chainB, latestHeight.GetRevisionHeight())
+			s.Require().NoError(err)
+		})
+
+		t.Run("get validators at latest height", func(t *testing.T) {
+			validators, err = s.GetValidatorSetByHeight(ctx, chainB, latestHeight.GetRevisionHeight())
+			s.Require().NoError(err)
+		})
+
+		t.Run("extract validator private keys", func(t *testing.T) {
+			privateKeys := s.extractChainPrivateKeys(ctx, chainB)
+			for i, pv := range privateKeys {
+				pubKey, err := pv.GetPubKey()
+				s.Require().NoError(err)
+
+				validator := tmtypes.NewValidator(pubKey, validators[i].VotingPower)
+
+				validatorSet = append(validatorSet, validator)
+				signers = append(signers, pv)
+			}
+		})
+	})
+
+	t.Run("create malicious header", func(t *testing.T) {
+		valSet := tmtypes.NewValidatorSet(validatorSet)
+		maliciousHeader, err = createMaliciousTMHeader(chainB.Config().ChainID, int64(latestHeight.GetRevisionHeight()), trustedHeight,
```

Comment:
worth noting this test can fail if the validator set changes (due to staking unbond tx). The only reason this would occur is if the e2e library injects random validator staking tx's to fuzz the system. Not something we need to worry about but thought I'd point it out